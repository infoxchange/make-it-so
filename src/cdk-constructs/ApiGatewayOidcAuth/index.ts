import * as SST from "sst/constructs";
import path from "node:path";

type ConstructScope = ConstructorParameters<typeof SST.Auth>[0];
type ConstructId = ConstructorParameters<typeof SST.Auth>[1];

type Props = Omit<SST.AuthProps, "authenticator"> & {
  oidcIssuerUrl: string;
  oidcClientId: string;
  oidcScope: string;
};

export class ApiGatewayOidcAuth extends SST.Auth {
  constructor(scope: ConstructScope, id: ConstructId, props: Props) {
    super(scope, id, {
      ...props,
      authenticator: {
        handler: path.join(import.meta.dirname, "auth-route.handler"),
        environment: {
          OIDC_ISSUER_URL: props.oidcIssuerUrl,
          OIDC_CLIENT_ID: props.oidcClientId,
          OIDC_SCOPE: props.oidcScope,
        },
      },
    });
  }
}
