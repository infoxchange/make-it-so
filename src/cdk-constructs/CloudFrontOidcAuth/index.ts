import { Construct } from "constructs";
import SecretsManager from "aws-cdk-lib/aws-secretsmanager";
import CloudFront from "aws-cdk-lib/aws-cloudfront";
import CDK from "aws-cdk-lib";
import CdkCustomResources from "aws-cdk-lib/custom-resources";
import Lambda from "aws-cdk-lib/aws-lambda";
import * as SST from "sst/constructs";
import { Config as SSTInternalConfig } from "sst/config.js";
import CloudFrontOrigins from "aws-cdk-lib/aws-cloudfront-origins";
import { BaseSiteCdkDistributionProps } from "sst/constructs/BaseSite.js";
import path from "node:path";
import fs from "node:fs";

type ConstructScope = ConstructorParameters<typeof Construct>[0];
type ConstructId = ConstructorParameters<typeof Construct>[1];

type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

type Props = {
  oidcIssuerUrl: string;
  oidcClientId: string;
  oidcScope: string;
};

export class CloudFrontOidcAuth extends Construct {
  readonly oidcIssuerUrl: string;
  readonly oidcClientId: string;
  readonly oidcScope: string;
  readonly id: string;

  constructor(scope: ConstructScope, id: ConstructId, props: Props) {
    super(scope, id);
    this.oidcIssuerUrl = props.oidcIssuerUrl;
    this.oidcClientId = props.oidcClientId;
    this.oidcScope = props.oidcScope;
    this.id = id;
  }

  addToDistributionDefinition<
    DistributionProps extends BaseSiteCdkDistributionProps,
  >(
    scope: ConstructScope,
    {
      distributionDefinition,
      prefix = "/auth",
    }: { distributionDefinition: Mutable<DistributionProps>; prefix?: string },
  ) {
    const updatedDistributionDefinition = { ...distributionDefinition };
    const behaviourName = `${prefix.replace(/^\//g, "")}/*`;
    updatedDistributionDefinition.additionalBehaviors =
      updatedDistributionDefinition.additionalBehaviors
        ? { ...updatedDistributionDefinition.additionalBehaviors }
        : {};
    if (updatedDistributionDefinition.additionalBehaviors[behaviourName]) {
      throw new Error(
        `Behavior for prefix ${prefix} already exists in distribution definition`,
      );
    }

    const jwtSecret = new SecretsManager.Secret(this, `${this.id}JwtSecret`, {
      description: "JWT Signing Secret",
      generateSecretString: {
        passwordLength: 32,
        excludePunctuation: true,
        includeSpace: false,
        requireEachIncludedType: true,
      },
      // Secret is only used for sessions so it's safe to delete on stack removal
      removalPolicy: CDK.RemovalPolicy.DESTROY,
    });

    updatedDistributionDefinition.defaultBehavior = {
      ...updatedDistributionDefinition.defaultBehavior,
      functionAssociations: [
        ...(updatedDistributionDefinition.defaultBehavior
          ?.functionAssociations || []),
        this.getFunctionAssociation(scope, jwtSecret),
      ],
    };
    updatedDistributionDefinition.additionalBehaviors[behaviourName] =
      this.getAuthBehaviorOptions(scope, jwtSecret, prefix);
    return updatedDistributionDefinition;
  }

  private getFunctionAssociation(
    scope: ConstructScope,
    jwtSecret: SecretsManager.Secret,
  ): CloudFront.FunctionAssociation {
    const cfKeyValueStore = new CloudFront.KeyValueStore(
      scope,
      `${this.id}CFKeyValueStore`,
    );

    const kvStoreId = cfKeyValueStore.keyValueStoreId; // Your KV store ID
    const key = "jwt-secret";
    const kvsArn = `arn:aws:cloudfront::${CDK.Stack.of(this).account}:key-value-store/${kvStoreId}`;

    // Updating the KVM requires a valid ETag to be provided in the IfMatch parameter so we first must fetch the ETag
    const getEtag = new CdkCustomResources.AwsCustomResource(
      this,
      `${this.id}GetKVStoreEtag`,
      {
        installLatestAwsSdk: false, // No real benefit in our case for the cost of a longer execution time
        onUpdate: {
          // Since there's no onCreate, onUpdate will be called for CREATE events
          service: "@aws-sdk/client-cloudfront-keyvaluestore",
          action: "describeKeyValueStore",
          parameters: { KvsARN: kvsArn },
          // We include a timestamp in the physicalResourceId to ensure we fetch the latest etag on every update
          physicalResourceId: CdkCustomResources.PhysicalResourceId.of(
            `${kvStoreId}-etag-${Date.now()}`,
          ),
        },
        policy: CdkCustomResources.AwsCustomResourcePolicy.fromSdkCalls({
          resources: [kvsArn],
        }),
      },
    );
    const etag = getEtag.getResponseField("ETag");

    // An annoying limitation of CloudFormation is that it won't resolve dynamic references for secrets when
    // used as a parameter to a custom resource. To get around this we manually resolve it with another custom
    // resource. Note this won't result in the secret being exposed in CloudFormation templates but it will
    // be visible in the CloudWatch logs of the custom resource lambda. In our case that is acceptable.
    // https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/341
    const secretValue = new CdkCustomResources.AwsCustomResource(
      this,
      `${this.id}GetSecret`,
      {
        // There's no real benefit of fetching the latest sdk our case for the cost of a longer execution time
        installLatestAwsSdk: false,
        // Since there's no onCreate, onUpdate will be called for CREATE events
        onUpdate: {
          service: "@aws-sdk/client-secrets-manager",
          action: "getSecretValue",
          parameters: {
            SecretId: jwtSecret.secretArn,
          },
          // We include a timestamp in the physicalResourceId to ensure we fetch the latest secret value on every update
          physicalResourceId: CdkCustomResources.PhysicalResourceId.of(
            `${this.id}GetSecret-${Date.now()}`,
          ),
        },
        policy: CdkCustomResources.AwsCustomResourcePolicy.fromSdkCalls({
          resources: [jwtSecret.secretArn],
        }),
      },
    );

    // Now we can actually update the KVS with the secret value
    const putKeyValue = new CdkCustomResources.AwsCustomResource(
      this,
      `${this.id}PutKeyValue`,
      {
        installLatestAwsSdk: false, // No real benefit in our case for the cost of a longer execution time
        onUpdate: {
          // Since there's no onCreate, onUpdate will be called for CREATE events
          service: "@aws-sdk/client-cloudfront-keyvaluestore",
          action: "putKey",
          parameters: {
            KvsARN: kvsArn,
            Key: key,
            Value: secretValue.getResponseField("SecretString"),
            IfMatch: etag,
          },
          physicalResourceId: CdkCustomResources.PhysicalResourceId.of(
            `${kvStoreId}-${key}`,
          ),
        },
        policy: CdkCustomResources.AwsCustomResourcePolicy.fromSdkCalls({
          resources: [kvsArn],
        }),
      },
    );

    // putKey in the @aws-sdk/client-cloudfront-keyvaluestore package requires @aws-sdk/signature-v4-crt to be imported
    // as well. But AwsCustomResource doesn't give us direct access to the underlying Lambda function so we inject a
    // NODE_OPTIONS env var to import on start. At some point AwsCustomResource will presumably switch to a later node
    // version and we might need to update this to '--import=' instead of '--require='.
    const fn = putKeyValue.node.findChild("Provider");
    if (!(fn instanceof Lambda.SingletonFunction)) {
      throw new Error(
        "Could not find the underlying Lambda function of the AwsCustomResource",
      );
    }
    fn.addEnvironment("NODE_OPTIONS", "--require=@aws-sdk/signature-v4-crt");

    const authCheckFunction = new CloudFront.Function(
      scope,
      `${this.id}AuthCheckFunction`,
      {
        code: CloudFront.FunctionCode.fromInline(
          fs
            .readFileSync(
              path.join(import.meta.dirname, "auth-check.js"),
              "utf8",
            )
            .replace("__placeholder-for-jwt-secret-key__", key),
        ),
        runtime: CloudFront.FunctionRuntime.JS_2_0,
        keyValueStore: cfKeyValueStore,
      },
    );

    return {
      function: authCheckFunction,
      eventType: CloudFront.FunctionEventType.VIEWER_REQUEST,
    };
  }

  private getAuthBehaviorOptions(
    scope: ConstructScope,
    jwtSecret: SecretsManager.Secret,
    prefix: string,
  ): CloudFront.BehaviorOptions {
    const authRouteFunction = new SST.Function(
      scope,
      `${this.id}AuthRouteFunction`,
      {
        runtime: "nodejs20.x",
        handler: path.join(import.meta.dirname, "auth-route.handler"),
        environment: {
          OIDC_ISSUER_URL: this.oidcIssuerUrl,
          OIDC_CLIENT_ID: this.oidcClientId,
          OIDC_SCOPE: this.oidcScope,
          JWT_SECRET: jwtSecret.secretValue.toString(),
        },
      },
    );

    // authRouteFunction uses SST's AuthHandler construct which is normally run inside a lambda that's
    // created by SST's Auth construct. AuthHandler expects certain environment variables to be set
    // by the Auth construct so we have to set them ourselves here to keep it happy.
    const envVarName = SSTInternalConfig.envFor({
      type: "Auth",
      id: "id", // It seems like the env var will still be found no matter what this value is
      prop: "prefix",
    });
    authRouteFunction.addEnvironment(envVarName, prefix);

    const authRouteFunctionUrl = authRouteFunction.addFunctionUrl({
      authType: Lambda.FunctionUrlAuthType.NONE,
    });
    const forwardHostHeaderCfFunction = new CloudFront.Function(
      scope,
      `${this.id}ForwardHostHeaderFunction`,
      {
        code: CloudFront.FunctionCode.fromInline(`
        function handler(event) {
          const request = event.request;
          request.headers["x-forwarded-host"] = { value: request.headers.host.value };
          return request;
        }
      `),
        runtime: CloudFront.FunctionRuntime.JS_2_0,
      },
    );

    return {
      origin: new CloudFrontOrigins.HttpOrigin(
        CDK.Fn.parseDomainName(authRouteFunctionUrl.url),
      ),
      allowedMethods: CloudFront.AllowedMethods.ALLOW_ALL,
      cachePolicy: new CloudFront.CachePolicy(
        scope,
        `${this.id}AllowAllCookiesPolicy`,
        {
          cachePolicyName: "AllowAllCookiesPolicy",
          comment: "Cache policy that forwards all cookies",
          defaultTtl: CDK.Duration.seconds(1),
          minTtl: CDK.Duration.seconds(1),
          maxTtl: CDK.Duration.seconds(1),
          cookieBehavior: CloudFront.CacheCookieBehavior.all(),
          headerBehavior:
            CloudFront.CacheHeaderBehavior.allowList("X-Forwarded-Host"),
          queryStringBehavior: CloudFront.CacheQueryStringBehavior.all(),
          enableAcceptEncodingGzip: true,
          enableAcceptEncodingBrotli: true,
        },
      ),
      functionAssociations: [
        {
          function: forwardHostHeaderCfFunction,
          eventType: CloudFront.FunctionEventType.VIEWER_REQUEST,
        },
      ],
    };
  }
}
