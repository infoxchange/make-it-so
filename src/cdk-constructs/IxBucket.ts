import { Bucket } from "sst/constructs";
import { BucketEncryption } from "aws-cdk-lib/aws-s3";
import ixDeployConfig from "../deployConfig.js";

type ConstructScope = ConstructorParameters<typeof Bucket>[0];
type ConstructId = ConstructorParameters<typeof Bucket>[1];
type ConstructProps = Exclude<
  ConstructorParameters<typeof Bucket>[2],
  undefined
>;

export class IxBucket extends Bucket {
  constructor(
    scope: ConstructScope,
    id: ConstructId,
    props: ConstructProps = {},
  ) {
    const bucketProps: ConstructProps = {
      blockPublicACLs: true,
      ...props,
      cdk: {
        ...props.cdk,
        bucket: {
          enforceSSL: true,
          ...(ixDeployConfig.isIxDeploy
            ? { encryption: BucketEncryption.S3_MANAGED }
            : {}),
          ...props.cdk?.bucket,
        },
      },
    };

    super(scope, id, bucketProps);
  }
}
