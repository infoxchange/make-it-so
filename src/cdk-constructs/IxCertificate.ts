import { Construct } from "constructs";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { Certificate, ICertificate } from "aws-cdk-lib/aws-certificatemanager";
import { CustomResource } from "aws-cdk-lib";

type ConstructScope = ConstructorParameters<typeof Construct>[0];
type ConstructId = ConstructorParameters<typeof Construct>[1];

type Props = {
  domainName: string;
  subjectAlternativeNames?: string[];
  region?: string;
};

export class IxCertificate extends Construct {
  public acmCertificate: ICertificate;

  constructor(scope: ConstructScope, id: ConstructId, props: Props) {
    super(scope, id);
    this.acmCertificate = this.createCertificate(scope, id, props);
  }

  private createCertificate(
    scope: ConstructScope,
    id: ConstructId,
    props: Props,
  ): ICertificate {
    const certificateCreationLambdaArn =
      StringParameter.valueForStringParameter(
        scope,
        "/shared-services/acm/lambdaArn-v2",
      );
    const certificateCustomResource = new CustomResource(
      scope,
      "DomainCert-" + id,
      {
        resourceType: "Custom::CertIssuingLambda",
        serviceToken: certificateCreationLambdaArn,
        properties: {
          DomainName: props.domainName,
          ...(props.subjectAlternativeNames && {
            SubjectAlternativeNames: props.subjectAlternativeNames,
          }),
          ...(props.region && { CertificateIssuingRegion: props.region }),
        },
      },
    );
    return Certificate.fromCertificateArn(
      scope,
      id + "-AwsCertificate",
      certificateCustomResource.ref,
    );
  }
}
