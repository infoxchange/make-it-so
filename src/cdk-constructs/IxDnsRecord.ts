import { Construct } from "constructs";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { CustomResource } from "aws-cdk-lib";
import { remapKeys } from "../shared.js";

type ConstructScope = ConstructorParameters<typeof Construct>[0];
type ConstructId = ConstructorParameters<typeof Construct>[1];

type Props = {
  name: string;
  value: string;
  ttl?: number;
  hostedZoneId?: string;
} & (
  | {
      type: "A" | "CNAME" | "NS" | "SOA";
    }
  | {
      type: "ALIAS";
      aliasZoneId: string;
    }
);

export class IxDnsRecord extends Construct {
  constructor(scope: ConstructScope, id: ConstructId, props: Props) {
    super(scope, id);
    this.createDnsRecord(scope, id, props);
  }

  private createDnsRecord(
    scope: ConstructScope,
    id: ConstructId,
    constructProps: Props,
  ): void {
    const dnsRecordUpdaterLambdaArn = StringParameter.valueForStringParameter(
      scope,
      "/shared-services/route53/lambdaArn",
    );

    const lambdaProps = remapKeys(constructProps, {
      name: "RecordFQDN",
      value: "RecordValue",
      ttl: "RecordTTL",
      hostedZoneId: "HostedZoneId",
      type: "RecordType",
      aliasZoneId: "AliasZoneId",
    });

    new CustomResource(scope, id + "-CertificateCustomResource", {
      resourceType: "Custom::DNSRecordUpdaterLambda",
      serviceToken: dnsRecordUpdaterLambdaArn,
      properties: lambdaProps,
    });
  }
}
