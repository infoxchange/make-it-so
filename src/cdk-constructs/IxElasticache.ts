import { Construct } from "constructs";
import { CfnCacheCluster, CfnSubnetGroup } from "aws-cdk-lib/aws-elasticache";
import { IVpc, SecurityGroup, Peer, Port } from "aws-cdk-lib/aws-ec2";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { Stack } from "aws-cdk-lib";
import { IxVpcDetails } from "./IxVpcDetails.js";
import deployConfig from "../deployConfig.js";

type ConstructScope = ConstructorParameters<typeof Construct>[0];
type ConstructId = ConstructorParameters<typeof Construct>[1];
type CacheClusterProps = ConstructorParameters<typeof CfnCacheCluster>[2];

type Props = CacheClusterProps & {
  vpc?: IVpc;
  vpcSubnetIds?: string[];
};

export class IxElasticache extends Construct {
  cluster: CfnCacheCluster;

  connectionString: string;

  constructor(
    scope: ConstructScope,
    id: ConstructId,
    { vpc, vpcSubnetIds, ...elasticacheProps }: Props,
  ) {
    super(scope, id);

    // Setup cluster name
    if (!elasticacheProps.clusterName && deployConfig.isIxDeploy) {
      elasticacheProps.clusterName = `${Stack.of(this).stackName}`;
    }

    if (!elasticacheProps.cacheSubnetGroupName) {
      // Setup VPC
      if (!vpc && deployConfig.isIxDeploy) {
        const vpcDetails = new IxVpcDetails(scope, id + "-IxVpcDetails");
        vpc = vpcDetails.vpc;
      }

      // Setup VPC subnets
      if (vpc && !vpcSubnetIds) {
        if (deployConfig.isIxDeploy) {
          vpcSubnetIds = [1, 2, 3].map((subnetNum) =>
            StringParameter.valueForStringParameter(
              scope,
              `/vpc/subnet/private-${deployConfig.workloadGroup}/${subnetNum}/id`,
            ),
          );
        } else {
          vpcSubnetIds = vpc.privateSubnets.map((subnet) => subnet.subnetId);

          if (!vpcSubnetIds.length) {
            throw Error(`The vpc ${vpc.vpcId} has no private subnets.`);
          }
        }
      }
    }

    // Setup cluster security group
    if (vpc && vpcSubnetIds) {
      const subnetGroup = new CfnSubnetGroup(scope, "ElasticacheSubnetGroup", {
        subnetIds: vpcSubnetIds,
        description: "Subnet group for redis",
      });

      elasticacheProps.cacheSubnetGroupName = subnetGroup.ref;

      const namePrefix = elasticacheProps.clusterName
        ? elasticacheProps.clusterName
        : `${Stack.of(this).stackName}`;

      const securityGroup = new SecurityGroup(
        scope,
        "ElasticacheSecurityGroup",
        {
          vpc,
          allowAllOutbound: true,
          description: "Security group for Elasticache Cluster",
          securityGroupName: `${namePrefix}-elasticache`,
        },
      );

      for (const subnetIndex of [1, 2, 3]) {
        const cidr = StringParameter.valueForStringParameter(
          scope,
          `/vpc/subnet/private-${deployConfig.workloadGroup}/${subnetIndex}/cidr`,
        );
        securityGroup.addIngressRule(
          Peer.ipv4(cidr),
          Port.tcp(6379),
          `Allow access to Elasticache cluster from private ${deployConfig.workloadGroup} subnet`,
        );
      }

      elasticacheProps.vpcSecurityGroupIds = [securityGroup.securityGroupId];
    }

    // Create Redis Cluster
    this.cluster = new CfnCacheCluster(scope, "RedisCluster", elasticacheProps);

    if (elasticacheProps.engine === "redis") {
      this.connectionString = `redis://${this.cluster.attrRedisEndpointAddress}:${this.cluster.attrRedisEndpointPort}`;
    } else if (elasticacheProps.engine === "memcached") {
      this.connectionString = `${this.cluster.attrConfigurationEndpointAddress}:${this.cluster.attrConfigurationEndpointPort}`;
    } else {
      throw Error(`Unsupported engine: ${elasticacheProps.engine}`);
    }
  }
}
