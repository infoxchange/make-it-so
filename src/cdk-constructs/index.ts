export * from "./IxVpcDetails.js";
export * from "./IxCertificate.js";
export * from "./IxDnsRecord.js";
export * from "./IxNextjsSite.js";
export * from "./IxElasticache.js";
export * from "./IxApi.js";
export * from "./IxQuicksightWorkspace.js";
