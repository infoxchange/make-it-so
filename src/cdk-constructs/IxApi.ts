import { Api } from "sst/constructs";
import { IxCertificate } from "./IxCertificate.js";
import { IxDnsRecord } from "./IxDnsRecord.js";
import ixDeployConfig from "../deployConfig.js";
import { convertToBase62Hash } from "../lib/utils/hash.js";

type ConstructScope = ConstructorParameters<typeof Api>[0];
type ConstructId = ConstructorParameters<typeof Api>[1];
type ConstructProps = Exclude<ConstructorParameters<typeof Api>[2], undefined>;

export class IxApi extends Api {
  constructor(
    scope: ConstructScope,
    id: ConstructId,
    props: ConstructProps = {},
  ) {
    if (ixDeployConfig.isIxDeploy) {
      IxApi.setupCustomDomain(scope, id, props);
    }

    super(scope, id, props);

    if (ixDeployConfig.isIxDeploy) {
      this.createDnsRecords(scope);
    }
  }

  // This must be static because we need to call it in the constructor before super
  private static setupCustomDomain(
    scope: ConstructScope,
    id: ConstructId,
    props: ConstructProps,
  ): void {
    // Default to using domains names passed in by the pipeline as the custom domain
    if (ixDeployConfig.isIxDeploy && !("customDomain" in props)) {
      props.customDomain = {
        domainName: ixDeployConfig.siteDomains[0],
      };
    }

    this.setupCertificate(scope, id, props);
  }

  // This must be static because we need to call it in the constructor before super
  private static setupCertificate(
    scope: ConstructScope,
    id: ConstructId,
    props: ConstructProps,
  ): void {
    if (!props?.customDomain) return;

    if (typeof props.customDomain === "string") {
      props.customDomain = { domainName: props.customDomain };
    }

    const domainName = props.customDomain.domainName;
    if (domainName) {
      const domainCert = new IxCertificate(scope, id + "-IxCertificate", {
        domainName,
        region: "ap-southeast-2", // API Gateway wants southeast-2.
      });
      props.customDomain.isExternalDomain = true;
      props.customDomain.cdk = props.customDomain.cdk ?? {};
      props.customDomain.cdk.certificate = domainCert.acmCertificate;
    }
  }

  private createDnsRecords(scope: ConstructScope) {
    if (this.cdk.domainName?.name && this.cdk.domainName?.regionalDomainName) {
      const domainNameLogicalId = convertToBase62Hash(this.cdk.domainName.name);

      // API Gateway has a separate domain for using with a CNAME (regionalDomainName)
      new IxDnsRecord(scope, `DnsRecord-${domainNameLogicalId}`, {
        type: "CNAME",
        name: this.cdk.domainName.name,
        value: this.cdk.domainName?.regionalDomainName,
        ttl: 900,
      });
    }
  }
}
