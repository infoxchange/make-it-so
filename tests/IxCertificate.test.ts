import { test, expect } from "vitest";

import { Template } from "aws-cdk-lib/assertions";
import * as cdk from "aws-cdk-lib";
import { IxCertificate } from "../src/cdk-constructs";

test("basic example - calls custom resource lambda with correct props", () => {
  const stack = new cdk.Stack();
  new IxCertificate(stack, "IxVpcDetails", {
    domainName: "example.com",
  });
  const template = Template.fromStack(stack);

  expect(template.findResources("Custom::CertIssuingLambda")).toMatchSnapshot();
});

test("full example - calls custom resource lambda with correct props", () => {
  const stack = new cdk.Stack();
  new IxCertificate(stack, "IxVpcDetails", {
    domainName: "example.com",
    subjectAlternativeNames: ["other-domain.com", "another-domain.com"],
    region: "us-west-2",
  });
  const template = Template.fromStack(stack);

  expect(template.findResources("Custom::CertIssuingLambda")).toMatchSnapshot();
});
