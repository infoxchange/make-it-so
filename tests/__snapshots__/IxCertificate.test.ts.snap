// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic example - calls custom resource lambda with correct props 1`] = `
{
  "DomainCertIxVpcDetails": {
    "DeletionPolicy": "Delete",
    "Properties": {
      "DomainName": "example.com",
      "ServiceToken": {
        "Ref": "SsmParameterValuesharedservicesacmlambdaArnv2C96584B6F00A464EAD1953AFF4B05118Parameter",
      },
    },
    "Type": "Custom::CertIssuingLambda",
    "UpdateReplacePolicy": "Delete",
  },
}
`;

exports[`full example - calls custom resource lambda with correct props 1`] = `
{
  "DomainCertIxVpcDetails": {
    "DeletionPolicy": "Delete",
    "Properties": {
      "CertificateIssuingRegion": "us-west-2",
      "DomainName": "example.com",
      "ServiceToken": {
        "Ref": "SsmParameterValuesharedservicesacmlambdaArnv2C96584B6F00A464EAD1953AFF4B05118Parameter",
      },
      "SubjectAlternativeNames": [
        "other-domain.com",
        "another-domain.com",
      ],
    },
    "Type": "Custom::CertIssuingLambda",
    "UpdateReplacePolicy": "Delete",
  },
}
`;
